### PR TITLE
Rock 1.0.22 - batch two

### PR DESCRIPTION
**Pull request by thomasfanell**
- added AST printer
- fixed an error where the auto-generated `Enum values()` would crash if the enum elements are binary operations
- support function call with arguments on tuples
  - this is now supported: `(o1, o2, o3) someFunction(someArgument)`
- improved obfuscator

### CODE REVIEW

File: source/rock/frontend/AstPrinter.ooc [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    BaseType, BinaryOp, Block, BoolLiteral, CallChain, Cast, CharLiteral,
+    ClassDecl, CommaSequence, Comparison, Conditional, ControlStatement,
+    CoverDecl, Declaration, Dereference, Else, EnumDecl, Expression,
+    FloatLiteral, FlowControl, Foreach, FunctionCall, FunctionDecl,
+    FuncType, If, Import, Include, InterfaceDecl, InterfaceImpl, IntLiteral,
+    Line, Literal, Match, Module, NamespaceDecl, Node, NullLiteral, OperatorDecl,
+    Parenthesis, PropertyDecl, RangeLiteral, Return, SafeNavigation, Scope,
+    Statement, StringLiteral, StructLiteral, TemplateDef, Ternary, Try, Tuple,
+    Type, TypeDecl, TypeList, UnaryOp, Use, UseDef, VariableAccess, VariableDecl,
+    Version, Visitor, While]
```
Comment (marcusnaslund) [inline source/rock/frontend/AstPrinter.ooc]: Woah

Comment (thomasfanell) [inline source/rock/frontend/AstPrinter.ooc]: The more the merrier

Comment (marcusnaslund) [inline source/rock/frontend/AstPrinter.ooc]: You might consider `import EVERYTHING` as an addition for rock 1.0.23. :)

Comment (shamanas) [inline source/rock/frontend/AstPrinter.ooc]: Yeah, the AST being split into a file for each Node type leads to this kind of thing unfortunately :P

File: source/rock/middle/EnumDecl.ooc [hunk 2]
```diff
@@ -45,43 +45,39 @@ EnumDecl: class extends TypeDecl {
...
         result
     }
 
     generateValuesFunction: func -> FunctionDecl {
-        result := FunctionDecl new("values", nullToken)
+        result := FunctionDecl new("values", token)
         typeName := match lastElementValue {
             case intLiteral: IntLiteral => "Int"
             case floatLiteral: FloatLiteral => "Float"
+            // Probably a BinaryOp, so we use Int
```
Comment (marcusnaslund) [inline source/rock/middle/EnumDecl.ooc]: I like the "Probably" here, like "what could go wrong, let's call it an Int"

Comment (thomasfanell) [inline source/rock/middle/EnumDecl.ooc]: Thanks, I like it too!

File: CHANGELOG.md [hunk 3]
```diff
@@ -7,6 +7,7 @@ Summary:
   - rock now auto-generates count() and values() functions for enums.
   - Improved the obfuscator
   - Added an AST printer, see `rock --help` on how to use it.
+  - You can now call methods that take arguments on tuples
```
Comment (marcusnaslund) [inline CHANGELOG.md]: What does this mean?

Comment (thomasfanell) [inline CHANGELOG.md]: `(foo, bar) someFunction(someArgument)`, this did not work previously

Comment (marcusnaslund) [inline CHANGELOG.md]: aaah good

Comment (thomasfanell) [inline CHANGELOG.md]: Could I have phrased it better?

**Comment by thomasfanell** [comment]
@fredrikbryntesson

**Pull request merged**